### PR TITLE
FT: Support completeMPU, objectCopy replication

### PR DESCRIPTION
**Pull request by bennettbuchanan**
Update object MD replicationInfo field when using completeMPU and objectCopy APIs and replication applies to the object.

### CODE REVIEW

**Comment by bennettbuchanan** [comment]
@ironman-machine try

File: lib/api/completeMultipartUpload.js [hunk 1]
```diff
@@ -433,6 +434,11 @@ function completeMultipartUpload(authInfo, request, log, callback) {
                     metaStoreParams.versioning = options.versioning;
                     metaStoreParams.isNull = options.isNull;
                     metaStoreParams.nullVersionId = options.nullVersionId;
+                    const replicationInfo = getReplicationInfo(objectKey,
```
Comment (jonathan-gramain) [inline lib/api/completeMultipartUpload.js]: Since this new processing is not related to versioning, could we rename the `versioningPreprocessing` function to some more generic name?

File: tests/unit/api/objectReplicationMD.js [hunk 2]
```diff
@@ -114,6 +122,92 @@ function createBucketWithReplication(hasStorageClass) {
     });
 }
 
+// Initiate an MPU, put a part with the given body, and complete the MPU.
+function putMPU(key, body, cb) {
+    return async.waterfall([
```
Comment (rahulreddy) [inline tests/unit/api/objectReplicationMD.js]: Unit tests have to use dummy data, otherwise these might as well be a functional tests. You can setup the mpu bucket and part data in `metadata.keyMaps` and when you do completeMultipartUpload it would "just work".

File: lib/api/completeMultipartUpload.js (hunk 1 above)
Comment (bennettbuchanan) [inline lib/api/completeMultipartUpload.js]: Thank you. I think it would be appropriate to instead move this to the previous waterfall function.

File: tests/unit/api/objectReplicationMD.js (hunk 2 above)
Comment (bennettbuchanan) [inline tests/unit/api/objectReplicationMD.js]: Updated.

**Comment by bennettbuchanan** [comment]
@ironman-machine try

File: tests/unit/api/objectReplicationMD.js [hunk 3]
```diff
@@ -27,6 +24,7 @@ const authInfo = makeAuthInfo(canonicalID);
 const ownerID = authInfo.getCanonicalID();
 const namespace = 'default';
 const bucketName = 'source-bucket';
+const mpuShadowBucket = `${constants.mpuBucketPrefix + bucketName}`;
```
Comment (rahulreddy) [inline tests/unit/api/objectReplicationMD.js]: You mean 
```js
const mpuShadowBucket = `${constants.mpuBucketPrefix}${bucketName}`;
```

Comment (bennettbuchanan) [inline tests/unit/api/objectReplicationMD.js]: Updated.

**Comment by bennettbuchanan** [comment]
@ironman-machine try

**Comment by bennettbuchanan** [comment]
E2E passed on rebuild: http://ci.ironmann.io/gh/scality/Integration/13960.

File: lib/api/completeMultipartUpload.js [hunk 4]
```diff
@@ -406,6 +407,11 @@ function completeMultipartUpload(authInfo, request, log, callback) {
                 multipart: true,
                 log,
             };
+            const replicationInfo = getReplicationInfo(objectKey, destBucket,
```
Comment (rahulreddy) [inline lib/api/completeMultipartUpload.js]: since `getReplicationInfo` returns `undefined` if there is no replication config, you can avoid the conditional and set
```js
const metaStoreParams = {
     authInfo,
     objectKey,
     replicationInfo: getReplicationInfo(objectKey, destBucket, false, calculatedSize),
     ...
}
```

File: lib/api/completeMultipartUpload.js [hunk 5]
```diff
@@ -5,6 +5,7 @@ const { errors, versioning, s3middleware } = require('arsenal');
 
 const escapeForXml = s3middleware.escapeForXml;
 const { pushMetric } = require('../utapi/utilities');
+const getReplicationInfo = require('./apiUtils/object/replication');
```
Comment (rahulreddy) [inline lib/api/completeMultipartUpload.js]: I realized that the filename does not reflect the module name. It should be
```js
const getReplicationInfo = require('./apiUtils/object/getReplicationInfo');
```

File: lib/api/completeMultipartUpload.js (hunk 4 above)
Comment (bennettbuchanan) [inline lib/api/completeMultipartUpload.js]: I will update this. I had considered that, but thought it would be better to not assign a property as `undefined`.

File: lib/api/completeMultipartUpload.js (hunk 5 above)
Comment (bennettbuchanan) [inline lib/api/completeMultipartUpload.js]: Thank you for spotting this. Updated.

**Comment by bennettbuchanan** [comment]
@ironman-machine try

File: lib/api/completeMultipartUpload.js (hunk 4 above)
Comment (rahulreddy) [inline lib/api/completeMultipartUpload.js]: `undefined` doesn't show up when you stringify the object, so that's good.

**Comment by bennettbuchanan** [comment]
@ironman-machine try

**Comment by bennettbuchanan** [comment]
@ironman-machine try

**Comment by bennettbuchanan** [comment]
E2E passed.
